autoextract/aio/client.py

```diff
@@ -77,19 +172,31 @@ async def request_raw(query: Query,
     """
     if agg_stats is None:
         agg_stats = AggStats()  # dummy stats, to simplify code
-    post = _post_func(session)
-    post_kwargs = dict(
-        url=endpoint,
-        json=query_as_dict_list(query),
-        auth=aiohttp.BasicAuth(get_apikey(api_key)),
-        headers={'User-Agent': user_agent(aiohttp)},
+
+    # Keep state between executions/retries
+    request_processor = RequestProcessor(
+        query=query,
+        handle_query_errors=handle_retries and not skip_query_errors,
     )
+
+    post = _post_func(session)
+    auth = aiohttp.BasicAuth(get_apikey(api_key))
+    headers = {'User-Agent': user_agent(aiohttp)}
+
     response_stats = []
     start_global = time.perf_counter()
 
     async def request():
         stats = ResponseStats.create(start_global)
         agg_stats.n_attempts += 1
+
+        post_kwargs = dict(
+            url=endpoint,
+            json=request_processor.pending_queries,
```

Comment:
My understanding of the batch handling logic:

* Make a batch request
* If there are query-level errors, retry them all in a single batch request. If there is a retry time specified for some of the errors, use max.
* Repeat until there are no errors or retry stop condition is met.
* Return a list with all results, including successes and last received errors (if they're not resolved by retries).

Is it correct? If so, what do you think about adding a comment somewhere, explaining this logic? It is spread over retry policy, RequestProcessor and request_raw function, so it is not straight forward to understand the behavior just from reading code.

Reply:
I've tried to better explain the retry logic within this commit https://github.com/scrapinghub/scrapinghub-autoextract/pull/7/commits/bc6e032b412bef7d79e04e14e0c5e7f4a567b0b6